toolchain/check/handle_index.cpp

```diff
@@ -16,17 +21,93 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
   return true;
 }
 
+// Returns the argument values of the `IndexWith` interface, this corresponds to
+// the `SubscriptType` and the `ElementType`, if the class does not implement
+// the said interface this returns an empty array reference.
+static auto GetIndexWithArgs(Context& context, Parse::NodeId node_id,
```

Comment:
```suggestion
// TODO: Switch to using an associated type instead of a parameter for the `ElementType`.
static auto GetIndexWithArgs(Context& context, Parse::NodeId node_id,
```

We should capture the need to switch to an associated type somewhere, maybe here because it affects how it'll be called?